### PR TITLE
Multiple fixes for unit-test framework and unit-test

### PR DESCRIPTION
**Pull request by chtyim**
- (CDAP-8434) Allow unit-test to have multiple runs if allowed by program type
  - Also fix how stopAll works
- (CDAP-8462) Fix race condition in AuthorizationTest that causes flaky result

### CODE REVIEW

**Comment by chtyim** [comment]
Building http://builds.cask.co/browse/CDAP-RUT572

File: cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java [hunk 1]
```diff
@@ -118,14 +116,18 @@ public WorkerManager getWorkerManager(String workerName) {
...
-          appFabricClient.stopProgram(application.getNamespace(), programId.getApplication(),
-                                      programId.getProgram(), programId.getType());
+      ApplicationDetail appDetail = appFabricClient.getVersionedInfo(application);
+      for (ProgramRecord programRecord : appDetail.getPrograms()) {
+        try {
+          appFabricClient.stopProgram(application.getNamespace(), application.getApplication(),
+                                      appDetail.getAppVersion(), programRecord.getName(), programRecord.getType());
+        } catch (BadRequestException e) {
+          // Ignore this as this will be throw if the program is not running, which is fine as there could
+          // be program runs that was stopped by itself and still exists in the runningProcesses
```
Comment (anwar6953) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java]: this comment about `runningProcesses` seems inaccurate now.
It seems like we call `stopProgram` for every program in the app, even if it wasn't started or if its not even running

Comment (chtyim) [reply cdap-unit-test/src/main/java/co/cask/cdap/test/internal/DefaultApplicationManager.java]: updated comment.

**Comment by anwar6953** [comment]
LGTM

**Comment by chtyim** [comment]
Built passed already. The last commit is just updating a comment. Not gooding to rerun the build again.

**Pull request merged**